### PR TITLE
Optimize BoundingBox and fix inaccuracy

### LINKED ISSUE #1903: BoundingBox inaccuracy
### Description

```pc.BoundingBox``` stores two vectors: ```center``` and ```halfExtents```. The minimum and maximum values can be calculated from these values using ```getMin``` and ```getMax```. The problem with this is rounding errors. The currently only way to get the actual minimum and maximum is to loop over the vertexbuffer and calculate them manually.

To fix this problem ```pc.BoundingBox``` has to be remade to store the minimum and maximum values and calculate center and halfextents from them.

### Steps to Reproduce

Run this code in any project.
```js
let min = new pc.Vec3(0.3,1,1);
let max = new pc.Vec3(2.7,1,1);
let aabb = new pc.BoundingBox();
aabb.setMinMax(min, max);
console.log(aabb.getMin().x);
// output: 0.2999999999999998
console.log(aabb.getMin().x === min.x);
// output: false
```

### PR DESCRIPTION
**Pull request by aidinabedi**
Fixes #1903

PR includes the following:
- Fix documentation for all public `pc.BoundingBox` methods (`copy`, `clone`, `compute`, `setMinMax`).
- Optimize `pc.BoundingBox` for both `min`/`max` and `center`/`halfExtents` by always keeping in sync.
- Support arrays (of length 3) as paramters (in addition to `pc.Vec3`) in `pc.BoundingBox#setMinMax`.
- Use new optimized `pc.BoundingBox#setMinMax` in `glb-parser.js` and `json-model.js` parser.
- Add new properties `min` and `max` to `pc.BoundingBox` (without re-calcuating on every get).
- Deprecate old "getters" (`getMin` and `getMax`) in `pc.BoundingBox`, while keeping backward-compatibility.

**PLEASE NOTE:** The behavior of `pc.BoundingBox` is still exactly the same. Only the internals have been improved. This PR is fully backward-compatible.

----
I confirm I have signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

File: src/shape/bounding-box.js [hunk 1]
```diff
@@ -293,6 +334,12 @@ Object.assign(pc, function () {
             );
         },
 
+        /**
+         * @function
+         * @name pc.BoundingBox#compute
+         * @description Compute the size of the AABB to encapsulate all specified vertices.
+         * @param {pc.Vec3[]} vertices - The vertices used to compute the new size for the AABB.
```
Comment (willeastcott) [inline src/shape/bounding-box.js]: I believe `vertices` is just an Array or Float32Array.

File: src/shape/bounding-box.js [hunk 2]
```diff
@@ -200,29 +226,44 @@ Object.assign(pc, function () {
...
+         * @param {pc.Vec3|number[]} min - The minimum corner of the AABB. Can be a {@link pc.Vec3} or an array (of length 3).
+         * @param {pc.Vec3|number[]} max - The maximum corner of the AABB. Can be a {@link pc.Vec3} or an array (of length 3).
+         */
         setMinMax: function (min, max) {
-            this.center.add2(max, min).scale(0.5);
-            this.halfExtents.sub2(max, min).scale(0.5);
+            if (!Array.isArray(min)) this._min.copy(min);
+            else this._min.set(min[0], min[1], min[2]);
+            if (!Array.isArray(max)) this._max.copy(max);
+            else this._max.set(max[0], max[1], max[2]);
```
Comment (willeastcott) [inline src/shape/bounding-box.js]: As @Maksims suggests, please add braces around `if` and `else` statements for readability.

Comment (aidinabedi) [reply src/shape/bounding-box.js]: Done

File: src/shape/bounding-box.js (hunk 1 above)
Comment (aidinabedi) [reply src/shape/bounding-box.js]: My bad. Fixed.

**Comment by aidinabedi** [comment]
@Maksims Thank you for catching that. I've fixed both.

**Comment by willeastcott** [comment]
I do have one more comment. Doing the following is not valid:

```
bbox.min.x = 1;
bbox.min.y = 2;
bbox.min.z = 3;

bbox.max.x = 4;
bbox.max.y = 5;
bbox.max.z = 6;
```

Are we concerned about that? 🤔

**Comment by aidinabedi** [comment]
@willeastcott Good point. That can be a problem. Interesting problem. Not sure if there is a nice way to solve that.

**Comment by willeastcott** [comment]
Unless we somehow switch wholesale to storing min/max and expose them as vanilla properties and create getters (no setters) for center + extents?

**Comment by aidinabedi** [comment]
@willeastcott Correct, I'm not in favor of switching to min/max wholesale since that would break the API. Instead I'm thinking of closing this PR and opening a new one that just includes the jsdoc fixes and minor improvements that don't affect the API.

Thank you for catching that Will, I am embarrassed for not seeing it first. I should probably avoid working on so many PRs at the same time. Unfortunately, it's hard to find uninterrupted time to focus.

**Comment by willeastcott** [comment]
Don't be embarrassed! Your contributions to the PlayCanvas codebase are amazing! Keep them coming. 😄

**Pull request closed without merging**